_drafts/2023-08-15-debugging-android-devices.md

```diff
@@ -0,0 +1,274 @@
+
+You can find more details here in the [official documentation](https://source.android.com/docs/security/features/selinux/).
+
+## Diagnostic Tools
+
+Due to the complexity of Android-based devices, Google ships several tools to help diagnose what is happening on them.
+
+### Android Debug Bridge (ADB)
+
+This allows you to connect to your device over USB or the network, run commands, pull logs, install apps, reboot the device, and push/pull files. This is a feature-rich tool that supports several other commands and options but those are the most commonly used.
+
+### Bug Reports
+
+Bug Reports are huge dumps of data from the device that collects information from all the services running on the device and creates a snapshot of the device at the time it was triggered. These are very resource-intensive and often will cause the device to appear frozen so you don’t normally want to do this frequently to a customer’s device.
+
+Android's official documentation has dozens of short explanations on exploring different types of data contained within a Bug Report.
+
+### DropboxManager
+
+DropboxManager is for app specific crash logs. As mentioned above, this is where Java exceptions, tombstones, kernel panics, etc can be found as individual files as opposed to having to try and pull them out of logcat.
+
+### Batterystats and Battery Historian
+
+The Android documentation describes Batterystats as "a tool included in the Android framework that collects battery data on your device". After using the device for a bit, you can run `adb shell dumpsys batterystats > batterystats.txt` to dump the stats into a file. This is where Battery Historian comes into play. Battery Historian is a webapp that you can run locally to visualize the data created by Batterystats.
+
+To explore more information about how to use Batterystats and Battery Historian, visit Google's [documentation](https://developer.android.com/topic/performance/power/setup-battery-historian#BatteryHistorianCharts).
+
+While it is an older video, there's a great [YouTube video](https://www.youtube.com/watch?v=i3pw5aCAvOc) showing how to explore the visualizations created by Battery Historian.
+
+### Perfetto
+
+Perfetto is an open-source performance tracing and visualization tool primarily designed for analyzing and improving the performance of software systems, especially on the Android platform. It was initially developed by Google and is now maintained as a collaborative project by the Perfetto community.
```

Comment:
I think some links to documentation and a more fleshed out example would be good here.